src/hotspot/cpu/x86/macroAssembler_x86.hpp

```diff
@@ -730,6 +730,11 @@ class MacroAssembler: public Assembler {
   void andptr(Register dst, int32_t src);
   void andptr(Register src1, Register src2) { LP64_ONLY(andq(src1, src2)) NOT_LP64(andl(src1, src2)) ; }
 
+#ifdef _LP64
```

Comment:
Why is it x64-specific?

Reply:
I believe its needed.

TLDR.. Couple of check ins ago, I broke the 32-bit build, and that was the 'easy' fix..
```
/usr/bin/ld: /home/vpaprots/src/jdk/build/linux-x86-server-fastdebug/hotspot/variant-server/libjvm/objs/macroAssembler_x86.o: in function `MacroAssembler::andq(Register, AddressLiteral, Register)':
/home/vpaprots/src/jdk/make/hotspot/../../src/hotspot/cpu/x86/macroAssembler_x86.cpp:1224: undefined reference to `Assembler::andq(Register, Address)'
```
--------------
Long version? I just set up a 32-bit sandbox to explain better, but I still cannot explain why that symbol is missing. On 32-bit build:
```
nm ./build/linux-x86-server-fastdebug/hotspot/variant-server/libjvm/objs/assembler_x86.o | c++filt | grep -E "Assembler::and.\("
000045c0 T Assembler::andb(Address, Register)
00004850 T Assembler::andl(Address, Register)
00004780 T Assembler::andl(Address, int)
00004960 T Assembler::andl(Register, Address)
00004820 T Assembler::andl(Register, int)
00004a70 T Assembler::andl(Register, Register)
000046d0 T Assembler::andw(Register, Register)
```

That same object file, on a 64-bit build does have the andq symbol:
```
nm ./build/linux-x86_64-server-fastdebug/hotspot/variant-server/libjvm/objs/assembler_x86.o | c++filt | grep -E "Assembler::and.\("
000000000005d710 T Assembler::andb(Address, Register)
000000000005d880 T Assembler::andl(Address, Register)
0000000000059800 T Assembler::andl(Address, int)
000000000005d9e0 T Assembler::andl(Register, Address)
00000000000044e0 T Assembler::andl(Register, Register)
0000000000004440 T Assembler::andl(Register, int)
00000000000793d0 T Assembler::andq(Address, Register)
0000000000078f50 T Assembler::andq(Address, int)
00000000000791a0 T Assembler::andq(Register, Address)
0000000000079310 T Assembler::andq(Register, Register)
0000000000079120 T Assembler::andq(Register, int)
0000000000004380 T Assembler::andw(Register, Register)
````

The mystery is that `assembler_x86.hpp`  treats `andq` and `andl` the same, there are no `#ifdef _LP64`.. yet somehow `andq` symbol is missing. I started tracing with `LOG=trace` but that produces quite an overwhelming amount of output, so figure can ask for advice.

Reply:
Right, `addq` instructions are x64-specific. I was confused because `assembler_x86.hpp` doesn't declare them as such which is a bug.

Reply:
I am mystified at how it actually gets removed from the `assembler_x86.o` object on 32-bit.. The only reliable/portable way _would_ be with `#ifdef` but its not there.. so.. code-generation? `sed`-like preprocessing? Can one edit object files after the gcc ran? The build must be doing something clever!! Haven't seen it yet.. 

Whatever the trick is, `assembler_x86.hpp` gets it, but not `macroAssembler_x86.hpp`.

If it doesn't ring any bells, maybe I will spend some more time looking at the traces, maybe can figure out what the build script is doing to remove the symbol.

Reply:
It's not specific to `andq`: there's a huge `#ifdef` block around the definitions in `assembler_x86.hpp` (lines 12201 - 13773; and there's even a nested `#ifdef _LP64` (lines 13515-13585)!) , but declarations aren't guarded by `#ifdef _LP64`.

Reply:
Yeah, just got to about the same conclusion by looking at the preprocessor `-E` output.. its declared in the header, but not defined in the 'cpp' file.. One would think that that's a compile error, but its been more then a decade since I looked at the C++ spec; 'C++ compiler is always right'.

Reply:
Don't know that there is anything else for me to do here? `assembler_x86.hpp` `#ifdef _LP64` macros were there before (and it not 'that wrong' or if a better/clean fix exists). `macroAssembler_x86.hpp` has to mirror that with `andq`.

(Just going through all the comments, making sure they have been addressed.) 

PS: In general I get worried about having macros changing object layout, but that's 'water under the bridge' and 64-bit seems big enough reason to have different layout. But its always 'entertaining debugging session' when offset of `a.f` is different in `a.o` and `b.o`, because somebody forgot to define same macros for `b.c` compile command as for `a.c`..

Reply:
Leave it as is. It'll be addressed separately.